Subscription endpoint docs

ikas opened the pull request:
## Summary

This PR refers to [GTC-219](https://gfw.atlassian.net/browse/GTC-219). It rewrites the **subscriptions** section with up-to-date documentation.

## Testing instructions

If you feel like deep-diving into the contents of the PR, follow the steps below:

1. Clone the documentation repository and follow the [installation instructions on the README](https://github.com/resource-watch/doc-api#installation-and-usage).
2. Checkout the branch of this PR: `docs/subscription-endpoints`
3. Read the revamped endpoints section for **subscriptions**.

ethanroday commented on source/includes/api/_subscriptions.md:
```diff
@@ -1,704 +1,811 @@
-<aside class="notice">
-Remember — All subscription endpoints need to be authenticated.
-</aside>
+A subscription allows you to get notified of updates on a datasets' data. We strongly recommend that you read the [dataset concept](#dataset) and [dataset endpoints](#dataset6) sections before proceeding.
```
Could be useful here to include a couple of examples of why you might want subscriptions.

ikas replied:
Added a couple of examples of use cases for subscriptions, such as subscribing to deforestation alerts in the Amazon forest or track fire alerts in the user's area of residence. Hope this makes it clearer :)

ethanroday commented on source/includes/api/_subscriptions.md:
````diff
@@ -1,704 +1,811 @@
 ```
 
-Inside the `subscribable` object, one (or several) sub-objects can be declared. In the provided example, the `test_subscription` is provided, with both a `dataQuery` and a `subscriptionQuery`. The former will be evaluated as the subscription's content, while the latter will be evaluated to check if a subscription has changed since the last update. Both queries can contain two special keywords: {b ...
+Before we go into the details of managing subscriptions, it's important to understand that, while you can create a subscription for *any* dataset, some conditions must be met by the dataset for its corresponding subscriptions to be functional and actually work as described here.
````
Assuming I understand correctly, we should create a backlog item to return error responses if a dataset is not configured for subscriptions. @ikas can you confirm that this would make sense?

ikas replied:
@ethanroday yes, we should!

ethanroday commented on source/includes/api/_subscriptions.md:
```diff
@@ -1,704 +1,811 @@
+env         | The environment to which the subscription is associated. | String | 'production'
 
-### From country and region
+**Deprecation notice:** the default value for the `application` filter (currently, `gfw`) will be removed and the `application` filter will then have no default value. We recommend reviewing your application to ensure you set and load the correct `application` explicitly.
```
Will be removed when?

ikas replied:
We actually have no date for removal, though I agree we should plan to remove this. The main intention here is to prevent users from trusting that the `application` filter will always be defaulting to `gfw`.

Subscriptions were built this way from the start, and GFW is currently using it trusting that the application default will be `gfw` (thus, not providing the application param). If we remove this default value, we also need to consider doing a full-sweep to GFW to ensure that nothing gets broken there from removing this value.

ethanroday commented on source/includes/api/_subscriptions.md:
```diff
@@ -1,704 +1,811 @@
+-------------- | -------------- | --------------
+400            | ID is not valid| The id provided is not a valid subscription id.
+401            | Unauthorized   | No valid token was provided in the request headers.
+404            | Subscription not found | Either no subscription exists with the provided id, or the subscription with the id provided is not owned by the user who performed the request.
```
Technically if it's a different user's subscription that should be a 403, right?

ikas replied:
It should, yes, but the docs are reflecting the way it is currently implemented.

We should add another backlog item to return the correct error response when trying to access subscriptions that don't belong to the logged user.

ethanroday commented on source/includes/api/_subscriptions.md:
```diff
@@ -1,704 +1,811 @@
+--------------------- | :------------------------------------------------------------: | ----------------:
+`params.wdpaid`       | Id of the protected area in the WDPA                           | String
+
+#### Subscribing to a geostore
```
Confused about why I can subscribe to both an area and a geostore. I thought geostores contained areas?

ikas replied:
It's the other way around - an **area** may (or may not) contain a geostore :) 

An **area** is a way of defining a geographical region (i.e. you can create an **area** for a geostore, ISO code, WDPA id, etc.). So, I think the idea here is that you can define in the subscription which specific geographical region you are subscribing to (i.e. geostore, ISO, WDPA, etc), or you can just reference an **area**.

Don't know if this helps to clarify your doubts :)

ethanroday commented on source/includes/api/_subscriptions.md:
````diff
@@ -1,704 +1,811 @@
 ```
 
-To delete a subscription, use the following DELETE endpoint.
+For convenience, the RW API offers an additional endpoint to resend the confirmation email. In order to do this, you should perform a GET request to the `v1/subscriptions/:id/send_confirmation`. As with most of the other subscription endpoints, please keep in mind that you must be authenticated in order to use this endpoint.
````
What happens if the subscription is already confirmed and I accidentally request another confirmation email? Will it still come? Will the confirmation get switched back to false until I click the link?

ikas replied:
No changes are performed to the subscription when triggering another confirmation email (i.e. if the subscription was already confirmed, it stays that way). 

The email will always be sent, regardless of the status of the subscription (confirmed or not).

I will add this info to the endpoint docs, thanks for pointing this out :)

ethanroday commented on source/includes/api/_subscriptions.md:
````diff
@@ -1,704 +1,811 @@
+```
 
-### General subscription statistics
+**Note: unsubscribing is equivalent to deleting the subscription**.
````
Just functionally, you mean, or does it actually delete the subscription?

ikas replied:
It actually fully deletes the subscription - should we try to make it clearer?

ethanroday replied:
Would just make it a little more direct - just say, "When you unsubscribe, the underlying subscription resource is deleted."

ikas replied:
done :+1:

ikas commented:
@ethanroday thanks for your feedback :+1: 

I have followed your advice and moved the `Subscription lifecycle` section over to the beginning of the docs - which I agree it leads to a better overall understanding of subscriptions.

Regarding your questions, I have addressed them all in the docs per se, but I'll paste here the answers just in case!

**How do the keys in subscribable work? Where do I reference them?**
The keys in the subscribable object are just for organizational purposes - i.e. the `subscribable` field in the dataset could just simply be an array of objects. The keys allow the end-user to give a name for "which data" the query is looking at, but they are not used in any way from the subscription's perspective.

**How are the dataQuery and subscriptionQuery structured?**
`dataQuery` is the subscription's content, and it will be used to build the email notification that is sent to the end-user. `subscriptionQuery` is used to evaluate if a subscription has changed since the last update. This query should return a single row with the count of updated items since the last update check. If the value returned by this query is greater than 0, then all the subscriptions that reference this dataset will receive a notification.

**Can I have multiple emails on the same subscription?**
The same subscription can subscribe to multiple datasets (both `datasets` and `datasetsQuery` fields of the subscription are arrays). But at the moment, the same subscription can't have multiple emails.

**Why is datasets a list? Can I subscribe to multiple datasets in the same subscription?**
Yes, you can subscribe to multiple datasets in the same subscription, and that is the reason why datasets is a list :)

**What's the datasetsQuery param?**
`datasetsQuery` is an alternative way of creating subscriptions - you can either provide the dataset ids in the `datasets` array field of the subscription, or you can provide them in the `datasetsQuery`. Using `datasetsQuery` has the advantage of being able to define a different threshold for sending the email (i.e. with `datasets`, the email is sent if the `subscriptionQuery` returns a count greater than 0 - with `datasetsQuery`, the email is sent if the `subscriptionQuery` returns a count greater than the threshold defined in the `datasetsQuery` object. This was not clear at all in the current docs, sorry for that, but should be clearer now!
